OCPBUGS-13226: Clarifying release notes for change in behavior for se… (#62197)
openshift/openshift-docs: release_notes/ocp-4-11-release-notes.adoc

[id="ocp-4-11-legacy-service-account"]
==== LegacyServiceAccountTokenNoAutoGeneration is on by default

> In previous releases, two service account token secrets were generated when a service account was created:

Reviewer: ```suggestion
In previous releases, 3 service account token secrets were generated when a service account was created:
```

> Prior to {product-title} 4.11, two service account token secrets were generated when a service account was created:

Author: Thanks, updating!

* A service account token secret for authenticating to the internal {product-title} registry
* A service account token secret for accessing the Kubernetes API

Starting with {product-title} 4.11, this second service account token secret for accessing the Kubernetes API is no longer created. This is because the `LegacyServiceAccountTokenNoAutoGeneration` upstream Kubernetes feature gate was enabled in Kubernetes 1.24, which stops the automatic generation of secret-based service account tokens for accessing the Kubernetes API. After upgrading to {product-title} 4.11, any existing service account token secrets are not deleted and continue to function.

[NOTE]
====
Do not rely on these automatically generated secrets for your own use; they might be removed in a future {product-title} release.
====

> If your workload needs a service account token, use the `TokenRequest` API to request a bound service account token. Tokens obtained from the `TokenRequest` API are more secure than service account token secrets, because they have a bounded lifetime and are not readable by other API clients. For more information, see xref:../authentication/bound-service-account-tokens.html#bound-sa-tokens-configuring_bound-service-account-tokens[Using bound service account tokens].

Reviewer 2: Only applications running outside a cluster would use the TokenRequest API explicitly. Should we mention that tokens are automatically bound to pods by via a projected volume? These use the API implicitly (kubelet calls the API and writes the resulting token to pods).

Author: @stlaz thanks for the clarification. Would this work?

```suggestion
If your workload needs a service account token, request a bound service account token by volume projection or by using the `TokenRequest` API directly. Tokens obtained from the `TokenRequest` API are more secure than service account token secrets, because they have a bounded lifetime and are not readable by other API clients. For more information, see xref:../authentication/bound-service-account-tokens.html#bound-sa-tokens-configuring_bound-service-account-tokens[Using bound service account tokens].
```

Author: (Updated the PR and preview per above suggestion, but let me know if I need to make any adjustments)

> If your workload needs a service account token, request a bound service account token by using volume projection or by using the `TokenRequest` API directly. Tokens obtained from the `TokenRequest` API are more secure than service account token secrets, because they have a bounded lifetime and are not readable by other API clients. For more information, see xref:../authentication/bound-service-account-tokens.html#bound-sa-tokens-configuring_bound-service-account-tokens[Using bound service account tokens].

Reviewer: > request a bound service account token by using volume projection

Users need not to do the request explicitly, because when a workload is created, it will automatedly create a bound service account token by using volume projection.

> by using the `TokenRequest` API directly

Instead of asking users to call the `TokenRequest` API directly, users could do it by command, the command calls  the `TokenRequest` API:
```shell
oc create token <serviceaccount name>
```

Author: @HanielZx what are you recommending we change this to then? Should we just more generically say something this?

```suggestion
If your workload needs a service account token, use bound service account tokens. Bound service account tokens are more secure than service account token secrets, because they have a bounded lifetime and are not readable by other API clients. For more information, see xref:../authentication/bound-service-account-tokens.html#bound-sa-tokens-configuring_bound-service-account-tokens[Using bound service account tokens].
```

Reviewer: ```suggestion
Each workload will be automatically injected with a projected volume to obtain a bound service account token. If your workload needs an additional service account token, add additional projected volume in your workload manifest. Bound service account tokens are more secure than service account token secrets, because they have a bounded lifetime and .... For more information, see xref:../authentication/bound-service-account-tokens.html#bound-sa-tokens-configuring_bound-service-account-tokens[Using bound service account tokens].
```
The part of "..." could be replaced by https://github.com/openshift/openshift-docs/pull/61541#discussion_r1250767436

Author: I ended up cutting out that whole section about why bound SA tokens are preferred over creating secrets, because that really isn't the important part of this release note entry (the main thing is to point out that we're not generating one of the secrets anymore). And we usually keep release notes shorter and link to elsewhere for more information. So the info on why they are preferred will exist (once my other PR is merged) in the "Using bound service account tokens" section that we link to.

So I think we are covered here - let me know if that sounds okay to you.

> You can also manually create a service account token secret to create a token, if the security exposure of a non-expiring token in a readable API object is acceptable to you. For more information, see xref:../nodes/pods/nodes-pods-secrets.html#nodes-pods-secrets-creating-sa_nodes-pods-secrets[Creating a service account token secret].

Reviewer 2: nit: stylistically I'd say "create ... to obtain a token" would work better

Author: Good catch, agreed
